File: getting_started/208.实现-trie-前缀树.py
#
# @lc app=leetcode.cn id=208 lang=python3
#
# [208] 实现 Trie (前缀树)
#

# @lc code=start
class Trie:

    def __init__(self):
        
        # value = ''
        self.next = []
        for i in range(26):
            self.next.append(None)
        self.end = False
            
            
    def s2i(self, s:str) -> int:
        
        return ord(s) - ord('a')
    
        
    def insert(self, word: str) -> None:
        
        node = self
        for s in word:
            i = self.s2i(s)
            if node.next[i] is None:
                node.next[i] = Trie()
            node = node.next[i]
        node.end = True

    def search(self, word: str) -> bool:
        
        node = self
        for s in word:
            i = self.s2i(s)
            if node.next[i] is None:
                return False
            node = node.next[i]
        if node.end:
            return True
        else:
            return False

    def startsWith(self, prefix: str) -> bool:
        
        node = self
        for s in prefix:
            i = self.s2i(s)
            if node.next[i] is None:
                return False
            node = node.next[i]

        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

File: getting_started/208.实现-trie-前缀树.cpp
/*
 * @lc app=leetcode.cn id=208 lang=cpp
 *
 * [208] 实现 Trie (前缀树)
 */

// @lc code=start
class Trie {
public:
    Trie() {
        next = new Trie*[26];
        for(int i = 0;i<26;i++)
        {
            next[i] = nullptr;
        }
        end = false;
    }
    
    int s2i(char c)
    {
        return int(c) - int('a');
    }

    void insert(string word) {
        
        Trie *node = this;
        int n = word.size();
        for(int i = 0;i<n;i++)
        {
            char c = word[i];
            int j = this->s2i(c);
            if(!node->next[j])
            {
                node->next[j] = new Trie();
            }
            node = node->next[j];
        }
        node->end = true;

    }
    
    bool search(string word) {

        Trie *node = this;
        int n = word.size();
        for(int i = 0;i<n;i++)
        {
            char c = word[i];
            int j = this->s2i(c);
            if(node->next[j])
            {
                node = node->next[j];
            }
            else
            {
                return false;
            }
        }
        if(node->end)
        {
            return true;
        }
        else
        {
            return false;
        }
        
    }
    
    bool startsWith(string prefix) {
        
        Trie *node = this;
        int n = prefix.size();
        for(int i = 0;i<n;i++)
        {
            char c = prefix[i];
            int j = this->s2i(c);
            if(node->next[j])
            {
                node = node->next[j];
            }
            else
            {
                return false;
            }
        }
        return true;
        
    }

private:
    Trie **next;
    bool end;
    bool flags[26];
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
// @lc code=end

